graphql/middleware.go

```diff
@@ -0,0 +1,36 @@
+package graphql
+
+import "context"
+
+type ComputationInput struct {
+	Id          string
+	Query       string
+	ParsedQuery *Query
+	Variables   map[string]interface{}
+	Ctx         context.Context
+	Previous    interface{}
+}
+
+type ComputationOutput struct {
+	Metadata map[string]interface{}
+	Current  interface{}
+	Error    error
+}
+
+type MiddlewareFunc func(input *ComputationInput, output *ComputationOutput, next NextFunc)
```

Comment:
Earlier version did this - I think it's reasonable, though it does impact when you can contribute to `Metadata`, i.e. you must wait until the base middleware (the executor) contributes `output.Metadata` before any middleware can write to it.

e.g.
```golang
c.Use(func(input *graphql.ComputationInput, output *graphql.ComputationOutput, next graphql.NextFunc) {
  requestId := middleware.RequestIdFromContext(input.Ctx)
  output.Metadata["requestId"] = requestId
  next(input, output)
})
```
would become:
```golang
c.Use(func(input *graphql.ComputationInput, next graphql.NextFunc) *graphql.ComputationOutput {
  requestId := middleware.RequestIdFromContext(input.Ctx)
  output := next(input, output)
  output.Metadata["requestId"] = requestId
})
```